src/librustc_mir/interpret/machine.rs

```diff
@@ -254,6 +254,15 @@ pub trait Machine<'mir, 'tcx>: Sized {
         kind: Option<MemoryKind<Self::MemoryKind>>,
     ) -> (Cow<'b, Allocation<Self::PointerTag, Self::AllocExtra>>, Self::PointerTag);
 
+    /// Called to notify the machine before a deallocation occurs
+    fn before_deallocation(
+        _memory_extra: &Self::MemoryExtra,
+        _id: AllocId,
+        _ptr: Pointer<Self::PointerTag>,
```

Comment:
What is the point of the `ptr` argument? The `id` already says what will be deallocated.

Reply:
I wasn't sure what was needed, so I just copied `AllocationExtra;:memory_deallocated`, if it's not necessary I can remove `ptr`.

Reply:
Oh... I had entirely forgotten about `AllocationExtra::memory_deallocated`. Since we already have that, I do not think we need a new machine hook for the diagnostic you are looking for. Sorry for the confusion.

Reply:
Ok, in that case how would I use `AllocationExtra::memory_deallocated`? I don't see an easy way to thread in `MemoryExtra` into `AllocationExtra`

Reply:
Hm, what do you need that for? Oh, to figure out which allocation is tracked.

Dang. Yeah okay let's add the hook.

Reply:
Ok, I'll go ahead and do that